File: streampark-common/src/main/scala/org/apache/streampark/common/util/HostsUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util

import org.apache.streampark.common.util.Implicits._

import io.netty.resolver.HostsFileParser

import java.net.InetAddress

import scala.collection.immutable.ListMap

object HostsUtils {

  /**
   * Get hosts info from system, entry of return Map [hostname -> ipv4]. The elements are sorted in
   * reverse order by the length of the hostname.
   */
  def getSortSystemHosts: ListMap[String, String] = {
    val ipMap =
      HostsFileParser.parse.inet4Entries.map(e => e._1 -> e._2.getHostAddress)
    ListMap(ipMap.toSeq.sortWith(_._1.length > _._1.length): _*)
  }

  /** Get hosts info from system, entry of return Map [hostname -> ipv4]. Scala api. */
  def getSystemHosts(excludeLocalHost: Boolean = false): Map[String, String] = {
    var map = HostsFileParser.parse.inet4Entries
      .map(e => e._1 -> e._2.getHostAddress)
      .toMap
    if (excludeLocalHost) {
      val localHostName = InetAddress.getLocalHost.getHostName
      map = map
        .filter(!_._1.equals("localhost"))
        .filter(!_._1.equals(localHostName))
        .filter(!_._2.equals("127.0.0.1"))
    }
    map
  }

  /** Get hosts info from system, entry of return Map [hostname -> ipv4]. Java api. */
  def getSystemHostsAsJava(excludeLocalHost: Boolean): JavaMap[String, String] =
    getSystemHosts(excludeLocalHost)

}

File: streampark-common/src/main/java/org/apache/streampark/common/util/HostsUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import io.netty.resolver.HostsFileParser;

import java.net.InetAddress;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Collectors;

public final class HostsUtils {

    private HostsUtils() {
    }

    public static Map<String, String> getSortSystemHosts() {
        Map<String, String> ipMap = new HashMap<>();
        try {
            HostsFileParser.parseSilently()
                .inet4Entries()
                .forEach((hostname, addr) -> ipMap.put(hostname, addr.getHostAddress()));
        } catch (Exception ignored) {
            // ignore
        }
        return ipMap.entrySet().stream()
            .sorted((a, b) -> Integer.compare(b.getKey().length(), a.getKey().length()))
            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (a, b) -> a, LinkedHashMap::new));
    }

    public static Map<String, String> getSystemHosts() {
        return getSystemHosts(false);
    }
    public static Map<String, String> getSystemHosts(boolean excludeLocalHost) {
        Map<String, String> map = new HashMap<>();
        try {
            HostsFileParser.parseSilently()
                .inet4Entries()
                .forEach((hostname, addr) -> map.put(hostname, addr.getHostAddress()));
        } catch (Exception ignored) {
            // ignore
        }
        if (excludeLocalHost) {
            try {
                String localHostName = InetAddress.getLocalHost().getHostName();
                map.entrySet().removeIf(e -> "localhost".equals(e.getKey()) || localHostName.equals(e.getKey())
                    || "127.0.0.1".equals(e.getValue()));
            } catch (Exception ignored) {
            }
        }
        return map;
    }

    public static Map<String, String> getSystemHostsAsJava(boolean excludeLocalHost) {
        return getSystemHosts(excludeLocalHost);
    }
}
